include/ensmallen_bits/agemoea/agemoea_impl.hpp

```diff
@@ -0,0 +1,806 @@
+  W = W + 1e-6;
+  std::vector<bool> selected(front.size());
+  arma::Col<ElemType> z(numObjectives, arma::fill::zeros);
+  arma::Row<ElemType> dists;
+  for (size_t i = 0; i < numObjectives; i++)
+  {
+    PointToLineDistance<MatType>(dists, calculatedObjectives, front, z, W.col(i));
+    for (size_t j = 0; j < front.size(); j++)
+      if (selected[j]){dists[j] = arma::datum::inf;}
+    indexes[i] = dists.index_min();
+    selected[dists.index_min()] = true;
+  }
+}
+
+//! Find the distance of a front from a line formed by two points.
+template <typename MatType>
+void AGEMOEA::PointToLineDistance(
+    arma::Row<typename MatType::elem_type>& distances,
+    std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives,
+    const std::vector<size_t>& front,
+    const arma::Col<typename MatType::elem_type>& pointA,
+    const arma::Col<typename MatType::elem_type>& pointB)
+{
+  typedef typename MatType::elem_type ElemType;
+  arma::Row<ElemType> distancesTemp(front.size());
+  arma::Col<ElemType> ba = pointB - pointA; 
+  arma::Col<ElemType> pa;
+
+  for (size_t i = 0; i < front.size(); i++)
+  {
+    size_t ind = front[i];
+ 
+    pa = (calculatedObjectives[ind] - pointA);
+    double t = arma::dot(pa, ba) / arma::dot(ba, ba);
+    distancesTemp[i] = arma::accu(arma::pow((pa - t * ba), 2));
+  }
+  distances = distancesTemp;
+}
+
+//! Sort population into Pareto fronts.
+template<typename MatType>
+inline void AGEMOEA::FastNonDominatedSort(
+    std::vector<std::vector<size_t> >& fronts,
+    std::vector<size_t>& ranks,
+    std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives)
+{
+  std::map<size_t, size_t> dominationCount;
+  std::map<size_t, std::set<size_t> > dominated;
+
+  // Reset and initialize fronts.
+  fronts.clear();
+  fronts.push_back(std::vector<size_t>());
+
+  for (size_t p = 0; p < populationSize; p++)
+  {
+    dominated[p] = std::set<size_t>();
+    dominationCount[p] = 0;
+
+    for (size_t q = 0; q < populationSize; q++)
+    {
+      if (Dominates<MatType>(calculatedObjectives, p, q))
+        dominated[p].insert(q);
+      else if (Dominates<MatType>(calculatedObjectives, q, p))
+        dominationCount[p] += 1;
+    }
+
+    if (dominationCount[p] == 0)
+    {
+      ranks[p] = 0;
+      fronts[0].push_back(p);
+    }
+  }
+
+  size_t i = 0;
+
+  while (!fronts[i].empty())
+  {
+    std::vector<size_t> nextFront;
+
+    for (size_t p: fronts[i])
+    {
+      for (size_t q: dominated[p])
+      {
+        dominationCount[q]--;
+
+        if (dominationCount[q] == 0)
+        {
+          ranks[q] = i + 1;
+          nextFront.push_back(q);
+        }
+      }
+    }
+
+    i++;
+    fronts.push_back(nextFront);
+  }
+  // Remove the empty final set.
+  fronts.pop_back();
+}
+
+//! Check if a candidate Pareto dominates another candidate.
+template<typename MatType>
+inline bool AGEMOEA::Dominates(
+    std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives,
```

Comment:
Why not just use `ColType` as a template parameter directly in this function and other functions?  It seems very awkward to do it like this because `MatType` is not actually used in these functions (at least I did not see any usage at a quick glance).

Reply:
I also felt the same but decided to keep it there since a similar pattern was followed `nsga2 `so I kept `agmoea` similar to it aiming for uniform implementation as `agemoea` is very similar instructure to `nsga2`.

I can do the changes but I recommend doing these changes for all the algorithms together in a later PR.

Reply:
It's a bit of a strange pattern, personally I think it would be a good idea to change it (it's clearer, and the way it currently is doesn't seem to provide any benefit).  Could be worth changing in NSGA2 too.

Reply:
Yes, I think its better if both are done together so I can pick it up after NSGA3 has been Implemented (Using the ColType ?) and then I can work on these two